```{r include=FALSE}
opts_chunk$set(echo=FALSE)
opts_chunk$set(warning=FALSE)
opts_chunk$set(fig.pos = 'H')

library(MotifBinner2)

if (!exists('result'))
{
  result <- all_results[[grep('primerSeqErr', names(all_results))[1]]]
}
if (class(result) != 'primerSeqErr')
{
  result <- all_results[[grep('primerSeqErr', names(all_results))[1]]]
}
```

# `r result$config$op_full_name`

Computes the sequencing error rates in the primer sequences.

Only reads where the scores of the alignment between the primer design and the reads is greater then 80% of the length of the primer design are considered for this calculation. If the score is less than this, then it is assumed that a process other than sequencing error is driving up the dissimilarity between the design and the read.

It is expected that the quality is extremely high in the primer region since they are at the starts of the reads.

Both forward and reverse reads must be looked at. They will be looked at seperately as well as pooled.

```{r, include=FALSE}
out <- NULL
for (data_source_name in sort(unique(result$metrics$primer_sequencing_stats$data_source)))
{
  out <- c(out, knit_child('primerSeqErr_internal.Rmd'))
}
```

```{r, results='asis'}
cat(paste(out, collapse = '\n'))
cat('\n\n')

kable_summary(result$summary)

timingTable(result)
```

```{r, results='asis'}
cat('\n\n---\n\n')
```

